app/models/run.rb

```diff
@@ -164,6 +164,26 @@ def publish_aging
     publish_age_every(1.day, 30)
   end
 
+  # Calculate the various statistical information about each segments history once in the database for the whole run
+  # instead of individually for each segment (N queries)
+  # Returns an array of SegmentHistories with the following attributes:
+  # segment_id, standard deviation, mean, median, and the 10th, 90th, 99th percentiles
+  def segment_history_stats
+    SegmentHistory.joins(segment: :run)
+                  .where(segment: {runs: {id: id}})
+                  .where.not(realtime_duration_ms: [0, nil])
+                  .group(:segment_id)
+                  .select('
+                    segment_id,
+                    STDDEV_POP(segment_histories.realtime_duration_ms) AS standard_deviation,
+                    AVG(segment_histories.realtime_duration_ms) AS mean,
+                    PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS median,
+                    PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile10,
+                    PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile90,
+                    PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile99
+                  ')
+  end
```

Comment:
Does there happen to be any way to put this method on `Segment` rather than on `Run`, while keeping the benefit of doing it all in one query per run?

Reply:
Does it make sense to put this on `Segment` when it generates the data for the whole run?  I was struggling to figure out where it should actually end up since it touches so many different parts.  Moving it around should be pretty easy though I think.

Reply:
I more mean is there a way to make e.g. `segment.stats` or whatever return that one segment's stats, but when called in a clever way (idk, like `segments.includes(:stats)`) we get all segments' stats loaded into memory and it is still run as one query. No idea if this is possible.

Also now that I'm thinking this through, it's feels a bit weird that this method returns an array of `SegmentHistory`s -- does that array include every segment history ever recorded in the run? If this method is coalescing data _about_ segment histories scoped to each segment, returning an array of `Segment`s seems more intuitive.

Reply:
Ok gotcha I see what you mean.  It's definitely possible, could have the find line that is current in the view extracted to the `Segment` model, and have that call the underlying function on the `Run` model where it is now.  That was each `Segment` can access it easily but still keep it on `Run` since it generates data for the whole run.

Is it worth converting it just to a normal hash maybe?  The data doesn't really feel like it fits into any of the existing models and it's basically a frankenstein model object the way that it currently is?

Reply:
Yeah frankenstein models do make me uneasy, a hash might be simpler to wrap my head around.